### PR TITLE
Added RPM package generation and upload

### LINKED ISSUE #36: Create and host RPM packages...
... for Fedora/Red Hat etc.

### PR DESCRIPTION
**Pull request by BClark09**
Requires testing to confirm the following before merge:

- [x] Automatic uploads in separate RPM specific repo
- [x] openHAB installs correctly using the [Repository Installation Paths](http://docs.openhab.org/installation/linux.html#file-locations)
- [x] openHAB upgrades correctly, preserving configurations files if changed by the user.
- [x] Upgrades only occur when a new version of the RPM package is released
- [x] Works the same on Fedora, CentOS and Red Hat
- [x] Does not change the functionality of the .Deb packages
- [x] Test addons and legacy-addons install

To help test, please use the following `/etc/yum.repos.d/openHAB.repo`

```
[openHAB]
name=openHAB
baseurl=https://openhab.jfrog.io/openhab/linuxpkg-testing-rpm/
enabled=1
gpgcheck=0
repo_gpgcheck=0
gpgkey=https://openhab.jfrog.io/openhab/api/gpg/key/public
```

install with:
```
yum install openhab2
yum install openhab2-addons
yum install openhab2-addons-legacy
```

Closes #36 

Signed-off-by: Ben Clark <ben@benjyc.uk>

### CODE REVIEW

**Comment by Stormwind99** [comment]
Error on Fedora 24:
```
[root@linus opt]# dnf install openhab2
Last metadata expiration check: 0:00:28 ago on Fri Feb 17 14:50:12 2017.
Error: nothing provides adduser needed by openhab2-2.1.0.20170217150518-1.noarch
(try to add '--allowerasing' to command line to replace conflicting packages)
```

adduser exists on the system, provided by shadow-utils package:
```
[root@linus opt]# which adduser
/usr/sbin/adduser
[root@linus opt]# rpm -q -f /usr/sbin/adduser
shadow-utils-4.2.1-8.fc24.i686
```

**Comment by BClark09** [comment]
Hmm, even getting passed that the openhab user doesn't exist because there's no RPM based addgroup. Bare with me for a bit.

**Comment by BClark09** [comment]
image

The latest version seems to work for testing.

**Comment by theoweiss** [comment]
@BClark09 I did the rpm packaging some weeks ago for a project at work. I've used the packaging of elasticsearch as template: 
You will need to add more deb/rpm specific handling into the script files as you can see here:
https://github.com/elastic/elasticsearch/tree/master/distribution/src/main/packaging/scripts
I suggest to stick with addgroup and adduser for debian like the elasticsearch guys did.

The script files are called with a different set of arguments depending if the executor is rpm or deb. Therefore you can distinguish  between rpm and deb installations.

**Comment by Stormwind99** [comment]
The yum repo packages still give me the adduser error - guessing the rpm in the repo has not been updated yet?

**Comment by BClark09** [comment]
Hmm, what happens when you 
````
dnf clean all
dnf update
```
?

**Comment by sipvoip** [comment]
Just a cosmetic thing, but you want way to rename it to openHAB rather then openHABRepo. Only issue I ran into was a LOT of warnings like:

warning: user openhab does not exist - using root
warning: group openhab does not exist - using root

Should not the .spec file create the user if it does not exist? Also, should think about signing the rpms. 

Was not able to install addons:

Transaction check error:
  file /usr/share/openhab2 from install of openhab2-addons-0:2.1.0.20170218132317-1.noarch conflicts with file from package openhab2-0:2.1.0.20170218132317-1.noarch

**Comment by BClark09** [comment]
Thanks for testing both, this is really helpful!

> Just a cosmetic thing, but you want way to rename it to openHAB rather then openHABRepo

Good point! Will change that when I come to writing it up in docs.

> warning: user openhab does not exist - using root
> warning: group openhab does not exist - using root

Right, I see that RPM and APT use different terms for [pre/post][inst/rm] and was trying to use the same ones. I've submitted another package to the repo to address this. Does this fix all the known errors (besides not being able to install addons)?

>  Also, should think about signing the rpms.

They will be signed with the same GPG key as the Deb files are when testing is complete.

**Comment by BClark09** [comment]
@theoweiss Have just seen your comment, sorry. Now that the basics are working I will follow the same style as you have showed.

**Comment by Stormwind99** [comment]
@BClark09 - dnf clean worked, but repos don't normally require that step to find newer packages.

The following worked and installled openhab2-2.1.0.20170218221203-1.noarch.rpm:
```
dnf clean
dnf install --nogpgcheck openhab2
```

**Comment by Stormwind99** [comment]
Error on installing openhab2-addons:
```
Error: Transaction check error:
  file /usr/share/openhab2 from install of openhab2-addons-0:2.1.0.20170218221203-1.noarch conflicts with file from package openhab2-0:2.1.0.20170218221203-1.noarch
```

**Comment by BClark09** [comment]
Hi @Stormwind99  and @sipvoip. I have completed the work on the rpm packages, and all known issues are resolved. Would you be able to test them for me? To do so you'll need to first 

```
yum remove openhab2
yum update
yum install openhab2
yum install openhab2-addons
```

if you want to simulate an upgrade, then you can pick the other version (has the same rpm setup as the latest) on the repo, simply by running the command:

```
yum install openhab2-2.1.0.20170225003526
```

Please let me know how you get on, I have tested a new install and upgrade of all packages on Fedora 25, but would like confirmation before I start moving to a full repo with GPG signing.

**Comment by BClark09** [comment]
@theoweiss, assuming that others are able to confirm my tests would you be happy for this PR (based on [this diff](https://github.com/openhab/openhab-linuxpkg/pull/37/files?w=1) to be merged?

**Comment by Stormwind99** [comment]
Both packages installed without error for me:
```
dnf remove openhab2
dnf update
dnf install --nogpgcheck openhab2
dnf install --nogpgcheck openhab2-addons
```

**Comment by sipvoip** [comment]
On Centos 7, I tested openhab2 first, and then addons, and also tested just
installed addons and having yum realize it needed to pull openhab2 also.
yum remove openhab2 also works, tho I don't know why anyone would want to
remove openHab. ;-)


><>
nathan stratton

File: resources/deb/control-runtime/postinst [hunk 1]
```diff
@@ -56,8 +56,8 @@ waitForStopFinished(){
 }
 
 case "$1" in
-	configure)
-		if [ -z $2 ] ; then
+	configure|1|2)
```
Comment (theoweiss) [inline resources/deb/control-runtime/postinst]: Could you add a comment with doc about what 1|2 args mean.

File: resources/deb/control-runtime/prerm [hunk 2]
```diff
@@ -22,11 +22,11 @@ stopOpenHAB() {
 case "$1" in
 	upgrade)
 		stopOpenHAB
-        exit 0
+		exit 0
 		;;
-	remove|deconfigure)
+	remove|deconfigure|0)
```
Comment (theoweiss) [inline resources/deb/control-runtime/prerm]: Could you add a comment with doc about the 0 arg.

File: resources/deb/control-runtime/postrm [hunk 3]
```diff
@@ -38,10 +38,29 @@ case "$1" in
 		rm -rf /etc/openhab2
 		exit 0
 		;;
-    upgrade)
-        removeCache
-        exit 0
-        ;;
+	0) # RPM has no purge, assume remove does this
```
Comment (theoweiss) [inline resources/deb/control-runtime/postrm]: Could you add a comment with doc about 0 arg.

**Comment by BClark09** [comment]
Thanks @Stormwind99 and @sipvoip for testing! And thanks @theoweiss for reviewing, I've added the documentation now so will merge and test a gpg signed version!

**Pull request merged**